File: 1671/solution.py
class Solution:
    def minimumMountainRemovals(self, nums: list[int]) -> int:
        lefts = [1] * len(nums)
        for i in range(len(nums)):
            for j in range(i):
                if nums[i] > nums[j]:
                    lefts[i] = max(1 + lefts[j], lefts[i])

        rights = [1] * len(nums)
        for i in range(len(nums) - 1, -1, -1):
            for j in range(len(nums) - 1, i, -1):
                if nums[i] > nums[j]:
                    rights[i] = max(1 + rights[j], rights[i])
        res = float("inf")
        for i in range(len(nums)):
            if lefts[i] > 1 and rights[i] > 1:
                res = min(res, len(nums) - (lefts[i] + rights[i] - 1))
        return res

File: 1671/solution.java
class Solution {
    public int minimumMountainRemovals(int[] nums) {
        int n = nums.length;
        int[] lefts = new int[n];
        int[] rights = new int[n];

        Arrays.fill(lefts, 1);
        Arrays.fill(rights, 1);

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                    lefts[i] = Math.max(1 + lefts[j], lefts[i]);
                }
            }
        }

        for (int i = n - 1; i >= 0; i--) {
            for (int j = n - 1; j > i; j--) {
                if (nums[i] > nums[j]) {
                    rights[i] = Math.max(1 + rights[j], rights[i]);
                }
            }
        }

        int res = Integer.MAX_VALUE;

        for (int i = 0; i < n; i++) {
            if (lefts[i] > 1 && rights[i] > 1) {
                res = Math.min(res, n - (lefts[i] + rights[i] - 1));
            }
        }

        return res;
    }
}

// time O(n^2)
// space O(n)